File: Scotch/Cassette.fs
namespace Scotch

open System
open System.IO
open Newtonsoft.Json
open Newtonsoft.Json.Converters

module Cassette =
    let locker = new Object()

    let ReadCassette (cassettePath: string) : HttpInteraction seq =
        if (not (File.Exists(cassettePath))) then
            Seq.empty
        else
            let jsonString = File.ReadAllText(cassettePath)
            let cassetteParseResult = JsonConvert.DeserializeObject<List<HttpInteraction>> (jsonString, new VersionConverter())
            List.toSeq cassetteParseResult

    let WriteCassette cassettePath (httpInteractions: HttpInteraction seq) =
            let serializedInteraction = JsonConvert.SerializeObject (Seq.toList httpInteractions, Formatting.Indented, new VersionConverter())
            File.WriteAllText (cassettePath, serializedInteraction.ToString())

    let UpdateInteraction cassettePath httpInteraction =
        lock locker ( fun () ->
            let existingInteractions = ReadCassette cassettePath
            let matchingIndex = existingInteractions
                                |> Seq.tryFindIndex (fun i -> Helpers.requestsMatch httpInteraction.Request i.Request)

            let newInteractions =
                match matchingIndex with
                | None -> Seq.append existingInteractions [|httpInteraction|]
                | Some i -> existingInteractions
                            |> Seq.mapi  (fun index item -> if (index = i) then httpInteraction else item)

            WriteCassette cassettePath newInteractions
       )

File: Scotch/Cassette.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Scotch;

public static class Cassette
{
    public static object locker = new();

    public static IEnumerable<HttpInteraction> ReadCassette(string cassettePath)
    {
        if (!File.Exists(cassettePath)) return new List<HttpInteraction>();

        var jsonString = File.ReadAllText(cassettePath);
        var cassetteParseResult = JsonConvert.DeserializeObject<List<HttpInteraction>>(jsonString, new VersionConverter());
        return cassetteParseResult;
    }

    public static void UpdateInteraction(string cassettePath, HttpInteraction httpInteraction)
    {
        lock (locker)
        {
            var existingInteractions = ReadCassette(cassettePath).ToList();
            var matchingIndex = existingInteractions.FindIndex(i => Helpers.RequestsMatch(httpInteraction.Request, i.Request));
            List<HttpInteraction> newInteractions;
            if (matchingIndex < 0)
            {
                newInteractions = existingInteractions.Append(httpInteraction).ToList();
            }
            else
            {
                newInteractions = existingInteractions.ToList();
                newInteractions[matchingIndex] = httpInteraction;
            }

            WriteCassette(cassettePath, newInteractions);
        }
    }

    public static void WriteCassette(string cassettePath, IEnumerable<HttpInteraction> httpInteraction)
    {
        var serializedInteraction = JsonConvert.SerializeObject(httpInteraction.ToList(), Formatting.Indented, new VersionConverter());
        File.WriteAllText(cassettePath, serializedInteraction.ToString());
    }
}
